### ISSUE TITLE
Investigate scrolling hitches reported by SRF

### ISSUE THREAD

**Issue by defagos** [SRGSSR/pillarbox-apple#1049]
**As an SRF app developer** I would like to be able to implement a TikTok-like vertical video navigation experience, but in which playback of items resumes when an item covers more than 50% of the screen, not when scrolling ends. At the moment minor hitches can be experienced when the 50% boundary between two neighboring items is  crossed.

## Hints

From @AndreKocherSRF:

We have a problem with video playback using the Pillarbox player on iOS. Every time `play()` and `becomeActive()` are called, there's a noticeable bottleneck. I tried running these two calls on `DispatchQueue.userInitiated`, but they need to be executed on the main thread. Therefore, I can't further analyze where exactly the bottleneck occurs in the Pillarbox player. If I don't call `play()`, but only load the video, there are no FPS drops. Perhaps you are already aware of this issue, or you might have an idea how to avoid these bottlenecks? We have been experiencing this issue a while ago but are currently remaking our video tab in the SRF News App.

The example video is in our `SnappableVideoPlayerView` of our repo https://github.com/mmz-srf/srf-apple in the develop branch. If you want to test this out you have to use the test or stage environment of our dev app.

https://github.com/user-attachments/assets/f3bb1d9f-620f-4b74-9303-93844ff2d1ad

The recording was made on an iPhone 14 Pro.

## Additional context

We performed an initial analysis. When transitioning between two items the app implements three behaviors involving Pillarbox:

- `play()` is called to resume playback for the new current item.
- `becomeActive()` is called for system integration.
- Audio session settings are updated.

As described by @AndreKocherSRF seems that `play()` is the main reason why the hitch occurs. What we have to determine is whether some work performed by Pillarbox makes the frame deadline being missed or if this occurs at the rendering server level, whether we schedule too many updates or for some other reasons.

## Literature

- https://developer.apple.com/documentation/xcode/understanding-user-interface-responsiveness
- https://developer.apple.com/documentation/xcode/understanding-hangs-in-your-app
- https://developer.apple.com/documentation/xcode/improving-app-responsiveness
- https://developer.apple.com/tutorials/Instruments/getting-started-with-hang-analysis
- https://developer.apple.com/documentation/xcode/diagnosing-performance-issues-early
- https://developer.apple.com/documentation/xcode/analyzing-the-performance-of-your-shipping-app

## WWDC

- https://developer.apple.com/videos/play/tech-talks/10855
- https://developer.apple.com/videos/play/tech-talks/10856
- https://developer.apple.com/videos/play/tech-talks/10857
- https://developer.apple.com/videos/play/wwdc2020/10077
- https://developer.apple.com/videos/play/wwdc2022/10082
- https://developer.apple.com/videos/play/wwdc2022/110350
- https://developer.apple.com/videos/play/wwdc2023/10248
- https://devstreaming-cdn.apple.com/videos/wwdc/2016/418kmcxcwg4psxzsjuy/418/418_hd_using_time_profiler_in_instruments.mp4?dl=1

## Acceptance criteria

- Hitches have been mitigated if possible.
- If not possible the hitch source has been documented and reasonable mitigations are proposed in the documentation (e.g. play when scrolling ends).

## Tasks

- [x] Reproduce in SRF News.
- [x] Reproduce in Pillarbox.
- [x] Eliminate potential issues on our side if any.
- [ ] Fix the issue:
    - [ ] Update implementation as needed to fix the issue (e.g. avoid property access that can lead to blocking the main thread, moving work off the main thread, etc).
    - [ ] Update the documentation if needed.
- [x] Fix performance issues related to user defaults observation:
    - [x] Optimize user defaults observation.
    - [x] Report issue to Commanders Act.
    - [x] Add warning to Core Business documentation.
    - [x] Create task to remove this warning once the Commanders Act SDK has been updated.
- [ ] Check analytics and metrics.

**Comment by defagos** [comment]
It seems that `avPlayerViewControllerPlaybackSpeedUpdatePublisher()` is contributing a bit since we perform a text search on the main thread. We can probably mitigate it by moving it to a background thread.

**Comment by defagos** [comment]
If the call stack text search, as well as `becomeActive()` and the audio session update are removed from SRF codebase, the scrolling becomes a lot smoother (except for the first item transition but maybe this is related to SRF code).

`becomeActive()` and the audio session update both seem to contribute a bit. We can try to mitigate the first since it belongs to our codebase. 

I tried to naively mitigate the audio session update as follows:

```swift
if session.category != category || session.mode != mode || session.categoryOptions != options {
    try? session.setCategory(category, mode: mode, options: options)
}
```

but it seems that reading settings has a similar footprint. I guess a mitigation should be applied higher up, possibly by performing this operation in a more surgical way, or by storing settings at the app level so that we don't read them directly from the audio session. Not perfect, though.

Anyway, this is just a preliminary analysis, until we profile the code more this should be considered with a pinch of salt.

**Comment by defagos** [comment]
I applied `becomeActive()` to the current story in our demo and noticed no hitch. Not to say this has no contribution but it should probably be small.

**Comment by defagos** [comment]
TL;DR, we should:

- Fix the rate publisher to perform text search in the background, or remove it entirely if possible.
- Optimize our trackers, especially the comScore one I guess.

This should solve all performance issues on our side.

SRF app developers should then perform more surgical audio session updates. With all those measures all noticed hitches should go away (though older devices might reveal some remaining optimization opportunities).

**Comment by defagos** [comment]
The call stack symbol extraction is itself costly and cannot itself be made from another thread (otherwise it would not deliver the expected stack). It is also fragile and was not working in all cases anyway on some devices (the search string was nowhere to be found).

For this reason we should really remove this nonsense. The corollary is that the effective playback speed will not be correctly updated when using the system video view, but I guess this is the price to pay for having something that works anyway.

**Comment by defagos** [comment]
Using Instruments and the _Time Profiler_ instrument we can gain some interesting insights. The strategy is inspired by the [old WWDC session](https://devstreaming-cdn.apple.com/videos/wwdc/2016/418kmcxcwg4psxzsjuy/418/418_hd_using_time_profiler_in_instruments.mp4?dl=1) listed above:

1. Zoom on the time profiler timeline in Instruments.
2. Slowly swipe between two neighboring stories. When a hitch is perceived we can see a > 100% CPU spike in the profiler. Pause execution in Instruments.
3. Locate the spike and drag an inspection window before it (and containing at most the spike itself). By playing with the window origin we can see potential heavy stack traces associated with the window before the spike.

## Commanders Act CPU spikes

If we apply the above approach to the code as it is on the branch (with call stack symbol extraction removed) we can observe > 100% CPU spikes. By adjusting the inspection window before these spikes we can always observe work related to a Commanders Act event being sent:

![Image](https://github.com/user-attachments/assets/f9bb29f0-97ec-4cc2-b7c1-c12569d8cd2c)

Two related operations in particular seem of interest:

- Insertion of information into the user defaults.
- Data serialization.

If we comment out the event sending call from our tracker implementation most issues disappear. Still we can have a look at further issue sources.

## Item replacement CPU spikes

**Comment by defagos** [comment]
An unrelated matter: Also checking main thread activity on our custom player layout vs. the system player layout shows we can likely move some time formatting work off the main thread to keep it less busy.

**Comment by defagos** [comment]
Some comments:

- Main-thread work currently made to sync player items with the `AVQueuePlayer` queue are not easy to optimize. We namely need to keep everything on the main thread so that the player state is always immediately consistent (e.g. just after adding an item, the item is visible in the player item list).
- The hitches we see depend on work performed by the player which might be diluted or concentrated depending on how an app implements its user experience. In our demo quite several tasks namely occur at the same time right in between stories (player creation, item sync with `AVPlayerItem` creation, tracking for the new / previous item), leading to small hitches. Some optimizations could be made to mitigate these, e.g. switching items after swiping to a new item (at the end of scrolling), disabling tracking or creating all players once and caching them. But for the moment we decided to keep our current implementation, with which we can later work on improvements ideas to the issue it reveals.
- Attempting to move Commanders Act work off the main thread also has an associated cost and would raise runtime warnings anyway, as some APIs would then be called from background threads. So there is no real issue with Commanders Act itself, it is only part of the small contributions that, added up, can sometimes lead to a hitch.

We still could get rid of some unnecessarily slow code, though. This is what can be submitted as PR.

For SRF apps the changes made should suffice to mitigate most of their hitches, except audio session updates. These can likely be optimized by carefully rewriting the code or moving the associated change to a background thread, though.

**Comment by defagos** [comment]
I did some more investigation work, focusing on waiting threads and short hitch thresholds. I investigated each tracker we use separately by commenting out others in URN items and playing with the stories demo.

# comScore

No meaningful contribution. Contrary to what is written above this tracker is not the culprit.

# Commanders Act

There is a non-negligible contribution due to events being sent. Maybe we can do something here.

<img width="1840" alt="Image" src="https://github.com/user-attachments/assets/f3b32665-dcc4-4c0b-a0c2-5d876d242dc2" />

# Metrics

We can see that there is a large contribution when disabling the tracker. This is something we can likely optimize.

<img width="1840" alt="Image" src="https://github.com/user-attachments/assets/0a582ccd-a7a6-41f0-b1c7-a09cf7c710a5" />

**Comment by defagos** [comment]
The user defaults observation that leads to issues on our end is related to our settings bundle, used for hidden features (currently access to the web demo only).

We could optimize our view hierarchy to have more local `SettingsBundle` observation, but several months after we introduced this feature we probably went a bit too far IMHO. We initially wanted to avoid testers confusing web and native players, but usually we ask them to test the content under _Lists_, not under _Showcase_. I doubt there is a risk of confusion nowadays. I therefore propose we remove the settings bundle entirely.

**Comment by defagos** [comment]
I can confirm the ideas mentioned above do fix hitches, even small ones.

I could verify that SRF News hitches are fixed as well. There remains a minor contribution due to the fact that `Player.becomeActive()` is called when moving between stories.

This minor contribution comes from `MPNowPlayingSession.becomeActiveIfPossible()`, we might attempt to mitigate it as well but apps should in general make a player instance active in a modal context, not really when swiping stories IMHO.

**Comment by defagos** [comment]
I reported the issue to [Commanders Act](https://github.com/CommandersAct/iOSV5/issues/19). Wait and see.

**Comment by defagos** [comment]
An update: Hitches are the combination of several sources that add up:

- Commanders Act: Because of how it misuses user defaults. [No fix in sight](https://github.com/CommandersAct/iOSV5/issues/19), though.
- Item `date()` and `accessLog()` access: These require RPCs to the media service daemon.

If removing Commanders Act tracker and replacing `date()` and `accessLog()` retrieval by simple `nil`s, all hitches disappear. Therefore:

- If we can mitigate RPCs in #1228, then hitches should almost disappear.
- Hopefully Commanders Act will deliver an update (or we will have a new provider by then), at which point all issues should have been addressed.
